src/arena.rs

```diff
@@ -221,6 +221,40 @@ impl<T> Arena<T> {
         }
     }
 
+    /// Get mutable references to two values inside this arena at once by
+    /// [`Index`], returning `None` if the corresponding index is not contained
+    /// in this arena.
+    ///
+    /// # Panics
+    ///
+    /// This function panics when the two indices are equal (having the same
+    /// slot number and generation).
+    pub fn get2_mut(&mut self, index1: Index, index2: Index) -> (Option<&mut T>, Option<&mut T>) {
+        if index1 == index2 {
+            panic!("Arena::get2_mut is called with two identical indices");
+        }
+
+        // Unsafe notes:
+        // - If `index1` and `index2` have different slot number, `item1` and
+        //   `item2` would point to different elements.
+        // - If `index1` and `index2` have the same slot number, only one could
+        //   be valid because there is only one valid generation number.
+        // - If `index1` and `index2` have the same slot number and the same
+        //   generation, this function will panic.
+        //
+        // Since `Vec::get_mut` will not reallocate, we can safely cast
+        // a mutable reference to an element to a pointer and back and remain
+        // valid.
+
+        let item1 = self.get_mut(index1);
+        let item1_bypass_borrow_checker = item1.map(|x| x as *mut T);
+
+        let item2 = self.get_mut(index2);
+        let item1 = unsafe { item1_bypass_borrow_checker.map(|x| x.as_mut().unwrap()) };
```

Comment:
```suggestion
        // Hold the first value in a pointer to sidestep the borrow checker
        let item1_ptr = self.get_mut(index1).map(|x| x as *mut T);

        let item2 = self.get_mut(index2);
        let item1 = unsafe { item1_ptr.map(|x| &mut *x) };
```

I think we can get away with this being a little more terse.